conf/machine/raspberrypi5.conf

```diff
@@ -0,0 +1,34 @@
+#@TYPE: Machine
+#@NAME: RaspberryPi 5 Development Board (64bit)
+#@DESCRIPTION: Machine configuration for the RaspberryPi 5 in 64 bits mode
+
+require conf/machine/include/arm/armv8-2a/tune-cortexa76.inc
+include conf/machine/include/rpi-base.inc
+
+MACHINE_FEATURES += "pci"
+MACHINE_EXTRA_RRECOMMENDS += "\
+    linux-firmware-rpidistro-bcm43455 \
+    bluez-firmware-rpidistro-bcm4345c0-hcd \
+    linux-firmware-rpidistro-bcm43456 \
+    bluez-firmware-rpidistro-bcm4345c5-hcd \
+"
+
+RPI_KERNEL_DEVICETREE = " \
+    broadcom/bcm2712-rpi-5-b.dtb \
+"
+
+SDIMG_KERNELIMAGE ?= "kernel_2712.img"
+SERIAL_CONSOLES ?= "115200;ttyAMA10"
+
+UBOOT_MACHINE = "rpi_arm64_config"
+
+VC4DTBO ?= "vc4-kms-v3d"
+
+# When u-boot is enabled we need to use the "Image" format and the "booti"
+# command to load the kernel
+KERNEL_IMAGETYPE_UBOOT ?= "Image"
+# "zImage" not supported on arm64 and ".gz" images not supported by bootloader yet
+KERNEL_IMAGETYPE_DIRECT ?= "Image"
+KERNEL_BOOTCMD ?= "booti"
+
+ARMSTUB ?= "armstub8-gic.bin"
```

Comment:
As per https://forums.raspberrypi.com/viewtopic.php?t=337129#p2018337 it looks like this is superfluous anyway. And the fact that it boots successfully should be proof the stub does it's job for the Pi5 too.

Reply:
> And the fact that it boots successfully should be proof the stub does it's job for the Pi5 too.

Well to be honest, it does not prove anything to me :) I will do some research and keep you informed.

Reply:
I am sorry, but could you help I was not able to make it worked, the boot hangs using the stub.

Reply:
TL;DR; It appears the stub does not work yet. You should drop it unless you know how to make it work.

---

I have enabled the debug on the second stage bootloader via [uart_2nstage](https://www.raspberrypi.com/documentation/computers/legacy_config_txt.html#uart_2ndstage), and here what I see:

Without the `armstub8-gic.bin` in `/boot`  and the kernel boots: here are the logs:

```
PCI2 init                                                                       
PCI2 reset                                                                      
PCIe scan 00001de4:00000001                                                     
RP1_CHIP_INFO 20001927                                                          
                                                                                
RPi: BOOTLOADER release VERSION:30de0ba5 DATE: 2023/10/30 TIME: 16:45:10        
BOOTMODE: 0x06 partition 0 build-ts BUILD_TIMESTAMP=1698684310 serial 64b59de4 boardrev d04170 stc 3987286
AON_RESET: 00000003 PM_RSTS 00001000                                            
usb_pd_init status 1                                                            
Boot mode: SD (01) order f4                                                     
SD HOST: 200000000 CTL0: 0x00800000 BUS: 400000 Hz actual: 390625 HZ div: 512 (256) status: 0x1fff0000 delay: 276
SD HOST: 200000000 CTL0: 0x00800f00 BUS: 400000 Hz actual: 390625 HZ div: 512 (256) status: 0x1fff0000 delay: 276
OCR c0ff8000 [394]                                                              
CID: 00123456534400000000220246030126                                           
CSD: 400e00325b590000761f7f800a400000                                           
SD: bus-width: 4 spec: 2 SCR: 0x02b58443 0x00000000                             
SD HOST: 200000000 CTL0: 0x00800f04 BUS: 50000000 Hz actual: 50000000 HZ div: 4 (2) status: 0x1fff0000 delay: 2
MBR: 0x00002000, 1048576 type: 0x0c                                             
MBR: 0x00102000,29908992 type: 0x83                                             
MBR: 0x00000000,       0 type: 0x00                                             
MBR: 0x00000000,       0 type: 0x00                                             
Trying partition: 0                                                             
type: 32 lba: 8192 'mkfs.fat' ' bootfs     ' clusters 261116 (4)                
rsc 32 fat-sectors 2040 root dir cluster 2 sectors 0 entries 0                  
FAT32 clusters 261116                                                           
[sdcard] autoboot.txt not found                                                 
Trying partition: 0                                                             
type: 32 lba: 8192 'mkfs.fat' ' bootfs     ' clusters 261116 (4)                
rsc 32 fat-sectors 2040 root dir cluster 2 sectors 0 entries 0                  
FAT32 clusters 261116                                                           
Read config.txt bytes     1229 hnd 0x137ec                                      
[sdcard] pieeprom.upd not found                                                 
usb_max_current_enable default 0 max-current 900                                
Read bcm2712-rpi-5-b.dtb bytes    73712 hnd 0x10417                             
dt-match: compatible: raspberrypi,5-model-b match: brcm,bcm2712                 
dt-match: compatible: brcm,bcm2712 match: brcm,bcm2712                          
PM_RSTS 00001000                                                                
Selecting USB low current limit                                                 
Read /config.txt bytes     1229 hnd 0x137ec                                     
Read /config.txt bytes     1229 hnd 0x137ec                                     
MESS:00:00:04.422951:0: Initial voltage 800000 temp 46622                       
MESS:00:00:04.622673:0: avs_2712: AVS pred 8526 852600 temp 46622               
MESS:00:00:04.625665:0: vpred 852 mV +0                                         
MESS:00:00:05.255394:0: FB framebuffer_swap 1                                   
MESS:00:00:05.284789:0: Select resolution HDMI0/2 hotplug 1 max_mode 2          
MESS:00:00:05.300596:0: HDMI0 edid block 0 offset 0                             
MESS:00:00:05.302363:0: 00ffffffffffff001e6d2e77dcf60000                        
MESS:00:00:05.307424:0: 0720010380502278eaca95a6554ea126                        
MESS:00:00:05.312486:0: 0f50542108007140818081c0a9c0b300                        
MESS:00:00:05.317548:0: d1c081000101e77c70a0d0a029503020                        
MESS:00:00:05.322610:0: 3a00204f3100001a9d6770a0d0a02250                        
MESS:00:00:05.327672:0: 30203a00204f3100001a000000fd0030                        
MESS:00:00:05.332734:0: 4b1e732b000a202020202020000000fc                        
MESS:00:00:05.337796:0: 004c472048445220575148440a200133                        
MESS:00:00:05.355235:0: HDMI0 edid block 1 offset 128                           
MESS:00:00:05.357179:0: 020346f1230907074b03041012131f5d                        
MESS:00:00:05.362241:0: 5e5f6061830100006d030c002000b83c                        
MESS:00:00:05.367302:0: 20006001020367d85dc401788003e30f                        
MESS:00:00:05.372364:0: 0030681a00000101304b00e305c000e6                        
MESS:00:00:05.377426:0: 0605015252567e4800e0a0381f404040                        
MESS:00:00:05.382488:0: 3a00204f31000018e4a770b8d1a02450                        
MESS:00:00:05.387550:0: 90608400204f31000000000000000000                        
MESS:00:00:05.392612:0: 00000000000000000000000000000063                        
MESS:00:00:05.397692:0: HDMI0: best-mode 2 (limit 2) 1920x1080 60 Hz CEA modes 18000d8000000000000000e003000000 extensions 1
MESS:00:00:05.408675:0: Select resolution HDMI1/2 hotplug 0 max_mode 2          
MESS:00:00:05.414985:0: FB0 disp 0 max-fb 2 1920x1080 stride 3840 base 0x3f800000
Loading 'initramfs_2712' to 0x00000000 offset 0x0                               
Read initramfs_2712 bytes 17117721 hnd 0x1173f                                  
MESS:00:00:06.175432:0: dtb_file 'bcm2712-rpi-5-b.dtb'                          
Loading 'bcm2712-rpi-5-b.dtb' to 0x00000000 offset 0x100                        
Read bcm2712-rpi-5-b.dtb bytes    73712 hnd 0x10417                             
Read /overlays/overlay_map.dtb bytes     4743 hnd 0x10563                       
Read /config.txt bytes     1229 hnd 0x137ec                                     
Read /overlays/vc4-kms-v3d-pi5.dtbo bytes     3330 hnd 0x10616                  
Read /cmdline.txt bytes      179 hnd 0x137eb                                    
BMD "armstub8-2712.bin" not found                                               
fs_open: 'armstub8-2712.bin'                                                    
Loading 'kernel_2712.img' to 0x00000000 offset 0x80000                          
Read kernel_2712.img bytes  8728960 hnd 0x10698                                 
PCI1 reset                                                                      
PCI2 reset                                                                      
MESS:00:00:08.354461:0: Starting OS 8354 ms                                     
MESS:00:00:08.358162:0: 00000040: -> 00000480                                   
MESS:00:00:08.359644:0: 00000030: -> 00100080                                   
MESS:00:00:08.363745:0: 00000034: -> 00100080                                   
MESS:00:00:08.367847:0: 00000038: -> 00100080                                   
MESS:00:00:08.371950:0: 0000003c: -> 00100080                                   
                                                                                
NOTICE:  BL31: v2.6(release):v2.6-239-g2a9ede0bd                                
NOTICE:  BL31: Built : 14:26:57, Jun 22 2023
```

The bootloader wants to load the file `armstub8-2712.bin`.

So I have added the armstub build for the pi4 and renamed it from `armstub8-gic.bin` to `armstub8-2712.bin` and the system seems to hang for the eternity (and even beyond, who knows):

```
PM_RSTS: 0x00001000                                                             
part 00000000 reset_info 00000000                                               
PMIC reset-event 00000000 rtc 00000001 alarm 00000000 enabled 0                 
uSD voltage 3.3V                                                                
Initialising SDRAM 'Micron' 32Gb x2 total-size: 64 Gbit 4267                    
DDR 4267 1 0 64 152                                                             
RP1_BOOT chip ID: 0x20001927                                                    
                                                                                
RP1_BOOT chip ID: 0x20001927                                                    
RP1_BOOT: fw size 25968                                                         
PCI2 init                                                                       
PCI2 reset                                                                      
PCIe scan 00001de4:00000001                                                     
RP1_CHIP_INFO 20001927                                                          
                                                                                
RPi: BOOTLOADER release VERSION:30de0ba5 DATE: 2023/10/30 TIME: 16:45:10        
BOOTMODE: 0x06 partition 0 build-ts BUILD_TIMESTAMP=1698684310 serial 64b59de4 boardrev d04170 stc 4011410
AON_RESET: 00000003 PM_RSTS 00001000                                            
usb_pd_init status 1                                                            
Boot mode: SD (01) order f4                                                     
SD HOST: 200000000 CTL0: 0x00800000 BUS: 400000 Hz actual: 390625 HZ div: 512 (256) status: 0x1fff0000 delay: 276
SD HOST: 200000000 CTL0: 0x00800f00 BUS: 400000 Hz actual: 390625 HZ div: 512 (256) status: 0x1fff0000 delay: 276
OCR c0ff8000 [483]                                                              
CID: 00123456534400000000220246030126                                           
CSD: 400e00325b590000761f7f800a400000                                           
SD: bus-width: 4 spec: 2 SCR: 0x02b58443 0x00000000                             
SD HOST: 200000000 CTL0: 0x00800f04 BUS: 50000000 Hz actual: 50000000 HZ div: 4 (2) status: 0x1fff0000 delay: 2
MBR: 0x00002000, 1048576 type: 0x0c                                             
MBR: 0x00102000,29908992 type: 0x83                                             
MBR: 0x00000000,       0 type: 0x00                                             
MBR: 0x00000000,       0 type: 0x00                                             
Trying partition: 0                                                             
type: 32 lba: 8192 'mkfs.fat' ' bootfs     ' clusters 261116 (4)                
rsc 32 fat-sectors 2040 root dir cluster 2 sectors 0 entries 0                  
FAT32 clusters 261116                                                           
[sdcard] autoboot.txt not found                                                 
Trying partition: 0                                                             
type: 32 lba: 8192 'mkfs.fat' ' bootfs     ' clusters 261116 (4)                
rsc 32 fat-sectors 2040 root dir cluster 2 sectors 0 entries 0                  
FAT32 clusters 261116                                                           
Read config.txt bytes     1228 hnd 0x137ed                                      
[sdcard] pieeprom.upd not found                                                 
usb_max_current_enable default 0 max-current 900                                
Read bcm2712-rpi-5-b.dtb bytes    73712 hnd 0x10417                             
dt-match: compatible: raspberrypi,5-model-b match: brcm,bcm2712                 
dt-match: compatible: brcm,bcm2712 match: brcm,bcm2712                          
PM_RSTS 00001000                                                                
Selecting USB low current limit                                                 
Read /config.txt bytes     1228 hnd 0x137ed                                     
Read /config.txt bytes     1228 hnd 0x137ed                                     
MESS:00:00:04.498861:0: Initial voltage 800000 temp 37280                       
MESS:00:00:04.698589:0: avs_2712: AVS pred 8489 848900 temp 37829               
MESS:00:00:04.701582:0: vpred 848 mV +0                                         
MESS:00:00:05.331420:0: FB framebuffer_swap 1                                   
MESS:00:00:05.360813:0: Select resolution HDMI0/2 hotplug 1 max_mode 2          
MESS:00:00:05.376623:0: HDMI0 edid block 0 offset 0                             
MESS:00:00:05.378389:0: 00ffffffffffff001e6d2e77dcf60000                        
MESS:00:00:05.383451:0: 0720010380502278eaca95a6554ea126                        
MESS:00:00:05.388513:0: 0f50542108007140818081c0a9c0b300                        
MESS:00:00:05.393575:0: d1c081000101e77c70a0d0a029503020                        
MESS:00:00:05.398637:0: 3a00204f3100001a9d6770a0d0a02250                        
MESS:00:00:05.403698:0: 30203a00204f3100001a000000fd0030                        
MESS:00:00:05.408760:0: 4b1e732b000a202020202020000000fc                        
MESS:00:00:05.413822:0: 004c472048445220575148440a200133                        
MESS:00:00:05.431261:0: HDMI0 edid block 1 offset 128                           
MESS:00:00:05.433205:0: 020346f1230907074b03041012131f5d                        
MESS:00:00:05.438267:0: 5e5f6061830100006d030c002000b83c                        
MESS:00:00:05.443329:0: 20006001020367d85dc401788003e30f                        
MESS:00:00:05.448391:0: 0030681a00000101304b00e305c000e6                        
MESS:00:00:05.453452:0: 0605015252567e4800e0a0381f404040                        
MESS:00:00:05.458514:0: 3a00204f31000018e4a770b8d1a02450                        
MESS:00:00:05.463576:0: 90608400204f31000000000000000000                        
MESS:00:00:05.468638:0: 00000000000000000000000000000063                        
MESS:00:00:05.473718:0: HDMI0: best-mode 2 (limit 2) 1920x1080 60 Hz CEA modes 18000d8000000000000000e003000000 extensions 1
MESS:00:00:05.484701:0: Select resolution HDMI1/2 hotplug 0 max_mode 2          
MESS:00:00:05.491011:0: FB0 disp 0 max-fb 2 1920x1080 stride 3840 base 0x3f800000
Loading 'initramfs_2712' to 0x00000000 offset 0x0                               
Read initramfs_2712 bytes 17117721 hnd 0x1173f                                  
MESS:00:00:06.249097:0: dtb_file 'bcm2712-rpi-5-b.dtb'                          
Loading 'bcm2712-rpi-5-b.dtb' to 0x00000000 offset 0x100                        
Read bcm2712-rpi-5-b.dtb bytes    73712 hnd 0x10417                             
Read /overlays/overlay_map.dtb bytes     4743 hnd 0x10563                       
Read /config.txt bytes     1228 hnd 0x137ed                                     
Read /overlays/vc4-kms-v3d-pi5.dtbo bytes     3330 hnd 0x10616                  
Read /cmdline.txt bytes      179 hnd 0x137eb                                    
Loading 'armstub8-2712.bin' to 0x00000000 offset 0x0                            
Read armstub8-2712.bin bytes      512 hnd 0x137ef                               
Loading 'kernel_2712.img' to 0x00000000 offset 0x80000                          
Read kernel_2712.img bytes  8728960 hnd 0x10698                                 
PCI1 reset                                                                      
PCI2 reset                                                                      
MESS:00:00:08.425572:0: Starting OS 8425 ms                                     
MESS:00:00:08.429273:0: 00000040: -> 00000480                                   
MESS:00:00:08.430754:0: 00000030: -> 00100080                                   
MESS:00:00:08.434856:0: 00000034: -> 00100080                                   
MESS:00:00:08.438958:0: 00000038: -> 00100080                                   
MESS:00:00:08.443060:0: 0000003c: -> 00100080
```

The `armstubs` [tree](https://github.com/raspberrypi/tools/blob/master/armstubs/armstub.S) has not received any updates since the pi5 has been released.

Note: the [Cortex-A72](https://developer.arm.com/documentation/100095/0003/Chunk709296044) (BCM2711) uses GICv3 while [Cortex-A76](https://developer.arm.com/documentation/100798/0400/xil1477667346785) (BCM2712) uses GICv4 and it may explain why it does not work, but again, I know nothing about that so do *NOT* trust my words.

Reply:
Did you have `enable_gic=1` in config.txt?

Reply:
Yes it had, unless I made a mistake.

Reply:
These were introduced (and inherited since) when quirks were needed for RPI4. I can't remember the specifics, but something was along the 64-bit support lines. I'm happy to have it dropped if it's not working at all. We will just have to make the stub recipe aware that this machine doesn't support it.